track can_connect state for mock pebble clients

rwcarlsen opened the pull request:
This introduces tracking of can_connect state for containers/pebble-clients in the test harness.  Container.can_connect uses the PebbleClient.get_system_info function to determine connectability.  I've added a _check_connection call to every mock client method (including get_system_info) that consults the testing backend for whether or not pebble is "available" to connect to. It raises a ConnectionError if we are simulating can_connect=False.

By default, can_connect is true for all containers - this is to help not break tests in other charms.  Calling harness.begin() causes can_connect to become false for all containers.  And calling harness.container_pebble_ready automatically switches can_connect to true before running the hook code.

I suspect introducing a change like this will break a lot of tests - probably not good.  Perhaps I should change harness.begin() to not set can_connect to false unless something like harness.simulate_can_connect() has been called.  On the other hand, dumping this change as implemented may help shake out bugs in charm code. Thoughts?

Fixes #691.

jameinel commented:
99% of all tests will be calling Harness.begin() because until you do, the
charm attribute isn't initialized. It would make sense for
'begin_with_initial_hooks' to set can_connect to false until appropriate,
but if we do that for begin() then the statement "By default, can_connect
is true" is not actually valid, because all the tests immediately set it to
False.

John
=:->

On Wed, Mar 16, 2022 at 3:56 PM Robert Carlsen ***@***.***>
wrote:

> This introduces tracking of can_connect state for
> containers/pebble-clients in the test harness. Container.can_connect uses
> the PebbleClient.get_system_info function to determine connectability. I've
> added a _check_connection call to every mock client method (including
> get_system_info) that consults the testing backend for whether or not
> pebble is "available" to connect to. It raises a ConnectionError if we are
> simulating can_connect=False.
>
> By default, can_connect is true for all containers - this is to help not
> break tests in other charms. Calling harness.begin() causes can_connect to
> become false for all containers. And calling harness.container_pebble_ready
> automatically switches can_connect to true before running the hook code.
>
> I suspect introducing a change like this will break a lot of tests -
> probably not good. Perhaps I should change harness.begin() to not set
> can_connect to false unless something like harness.simulate_can_connect()
> has been called. On the other hand, dumping this change as implemented may
> help shake out bugs in charm code. Thoughts?
>
> Fixes #691 <https://github.com/canonical/operator/issues/691>.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/canonical/operator/pull/724
> Commit Summary
>
>    - abc3f91
>    <https://github.com/canonical/operator/pull/724/commits/abc3f914c7a1239654cb11f414188d21cfef1623>
>    track can_connect state for mock pebble clients
>
> File Changes
>
> (2 files <https://github.com/canonical/operator/pull/724/files>)
>
>    - *M* ops/testing.py
>    <https://github.com/canonical/operator/pull/724/files#diff-37e8d80748bba21bb08c982f17b7879059bacc8d413b59da998a896fca18be33>
>    (51)
>    - *M* test/test_testing.py
>    <https://github.com/canonical/operator/pull/724/files#diff-ead5e89d7e5ab1974668fadf5d27df3da43cb2f63acf1a29977476a7ec6f8c12>
>    (3)
>
> Patch Links:
>
>    - https://github.com/canonical/operator/pull/724.patch
>    - https://github.com/canonical/operator/pull/724.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/canonical/operator/pull/724>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRQ7PPOWE7THM4VPILRMTVAI4IPANCNFSM5Q433G2Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

rwcarlsen commented:
Yeah - I did notice this :P  I think we will probably need to just keep can_connect true unless users explicitly say they want it to be auto through e.g.  `harness.simulate_can_connect()`.

pengale commented on ops/testing.py:
```diff
@@ -1044,8 +1053,22 @@ def __init__(self, unit_name, meta):
+
+    def _set_can_connect(self, pebble_client, val):
+        if pebble_client not in self._pebble_clients_can_connect:
+            msg = 'cannot set can_connect for the client - are you running a "real" pebble test?'
```
Do we want to just silently skip this operation in the cases where someone is running a real pebble test?

Real vs. fake pebble tests are distinguished by an environment variable, which means that folks are likely to use the same test suite for both, correct?

rwcarlsen replied:
The env var actually just affects testing code - nothing in the harness/framework.  I did look into this a bit recently - the idea of making it easy for charms to have their own real pebble tests, but currently, there really isn't any "working with both" code without explicitly writing pluggable fixture stuff like is in test_testing.py or test_pebble.py.  I think the hard error is probably good for now and we can relax it going forward if we start accomodatingcharm real pebble testing more explicitly.

rwcarlsen replied:
I've made some doc string improvements.

pengale commented:
> Yeah - I did notice this :P I think we will probably need to just keep can_connect true unless users explicitly say they want it to be auto through e.g. `harness.simulate_can_connect()`.

I like this change, and would make one addition: begin_with_initial_hooks should call harness.simulate_can_connect().

That gives folks a nice chance to discover the behavior when they are inspecting begin_with_initial_hooks -- something they're likely to do if they're trying to write their own similar test fixture. And it maximizes the chances that we are going to break tests that should be broken (because the charm author has not been careful), while keeping other tests functional.

rwcarlsen commented:
Sounds like people think this is worthwhile.  I'm going to go ahead and try to get it polished up and add some proper testing for it.  I'm really leaning toward not having `begin_with_initial_hooks()` auto-simulate can_connect by default and only do it if devs explicitly activate it.  I'm just not comfortable with all the ecosystem test breakage chaos that would ensue (charms on our CI here even fail with it - e.g. observability).

pengale commented:
> I'm just not comfortable with all the ecosystem test breakage chaos that would ensue (charms on our CI here even fail with it - e.g. observability).

Huh. I'm a little bit worried that this means that we have some actual issues with the ecosystem charms, if they're assuming that can_connect will succeed before it actually succeeds. :-/

rwcarlsen commented:
Yeah it's a two-edged sword.  I think having it simulated by default probably is safer in the long run but really seems  irresponsible short term for a framework.  How religious about not breaking people's tests (this generally won't affect production) should we be for point releases?

jnsgruk commented:
I think in the case of point releases, we really should be *quite* careful. I think we should try to get a better understanding of quite how many charm tests *might* break here (Just from the configured Github Actions I can see Prometheus and Alertmanager are broken).

@simskij @sed-i @balbirthomas -- can you investigate a bit here and see if this has caught an actual bug in those respective charms?

simskij commented:
One of the code paths that is failing in alertmanager currently is 

```
if not self.container.can_connect():
            self.unit.status = MaintenanceStatus("Waiting for pod startup to complete")
            return
# ...

layer_changed = self._update_layer(restart=False) 
```

which in turn invokes

```python
plan = self.container.get_plan()
```

which breaks with the following subsequent trace

```
File "/home/runner/work/operator/operator/.tox/unit/lib/python3.8/site-packages/ops/model.py", line 1192, in get_plan
    return self._pebble.get_plan()
  File "/home/runner/work/operator/operator/.tox/unit/lib/python3.8/site-packages/ops/testing.py", line 1502, in get_plan
    self._check_connection()
  File "/home/runner/work/operator/operator/.tox/unit/lib/python3.8/site-packages/ops/testing.py", line 1[317](https://github.com/canonical/operator/runs/5591343757?check_suite_focus=true#step:5:317), in _check_connection
    raise pebble.ConnectionError('cannot connect to pebble')
ops.pebble.ConnectionError: cannot connect to pebble
```

This is with `begin_with_initial_hooks`.

If the container returns that it can connect to pebble, but a subsequent call to `get_plan()` returns it can't, it sure sounds like we have a regression that needs to be addressed, especially if this change is to be scoped as a point release.

sed-i commented on ops/testing.py:
```diff
@@ -1275,9 +1310,15 @@ def __init__(self, backend: _TestingModelBackend):
+
+    def _check_connection(self):
+        if not self._backend._can_connect(self):
+            raise pebble.ConnectionError('cannot connect to pebble')
```
Alertmanager tests fail because of this `raise` here.
I wonder why this wrapper is needed instead of using `_can_connect` directly.

rwcarlsen replied:
Looks like there is some `@patch`ing causing the issue.

sed-i commented:
What's odd to me is that in alertmanager `can_connect` returns True, **followed by** `_check_connection -> ` raising.

The main guard is passed

```python
        if not self.container.can_connect():
            self.unit.status = MaintenanceStatus("Waiting for pod startup to complete")
            return
```
indicating that `get_system_info` didn't raise, but then a few lines down the file, `self.container.get_plan()` fails because `self._backend._can_connect -> False` and `_check_connection -> ` raises.

rwcarlsen commented:
The problem(s) that @simskij and @sed-i  brought up seem to be caused by the `@patch`ing of `get_system_info` on the pebble client - which obviously breaks the simulation of can_connect here since can_connect is based on it.  It seems like the code in this PR is working as expected and that enabling can_connect simulation by default would be far too disruptive for a point release.  Too much code does patching and/or assumes can_connect=true and pebble api calls work.

rwcarlsen commented:
To the deprecation point @jnsgruk suggested.  We could emit a deprecated warning if begin_with_initial_hooks runs without `simulate_can_connect` having been first called.  Not sure what conventions/practices we want to establish around deprecation, but at least for changes like this that improve correctness/robustness in testing generally it seems like it might be worthwhile.

rwcarlsen commented:
I added a warning that shows up in testing output.  I also added a way to globally turn activate can_connect simulation for all harness instances to facilitate conversion of entire test-suites without having to add in a simulate_can_connect call at *every* harness creation site.

sed-i commented:
@rwcarlsen I'm a bit confused by the latest changes:
1. Is the new can_connect feature (return True only after pebble ready) now disabled by default?
2. I expect the alertmanager tests to pass as-they-are with the new can_connect feature enabled. Is that the case now?

rwcarlsen commented:
`can_connect` simulation will be disabled by default opt-in for charms.

sed-i commented:
I don't understand this:

> caused by the `@patch`ing of `get_system_info` on the pebble client - which obviously breaks the simulation of can_connect here since can_connect is based on it.

Am I supposed to change my tests somehow for the new `can_connect` feature to work?

PietroPasotti commented on ops/testing.py:
```diff
@@ -76,6 +77,19 @@ class Harness(typing.Generic[CharmType]):
+    _global_simulate_can_connect = False
+
+    @staticmethod
+    def global_simulate_can_connect():
```
that value can be read from all users instead of having to be passed around by mutating class attributes and set by calling static methods

rwcarlsen replied:
Yeah - and that would allow us to consolidate future opt-in functionality vars all in one place.

rwcarlsen commented:
I've added some docs here: https://juju.is/docs/sdk/testing#heading--simulate-can-connect

rwcarlsen commented on ops/testing.py:
```diff
@@ -128,6 +169,14 @@ def begin(self) -> None:
+        # message below.  This warning was added 2022-03-22
+        if not self._backend._simulate_can_connect:
+            warnings.warn(
+                'Please call Harness.simulate_can_connect prior to calling Harness.begin.')
```
The message now points to a doc url with details explaining everything.

rwcarlsen commented:
@sed-i the alertmanager tests won't work with this new behavior because they patch the get_system_info function - which is how Container.can_connect determines the connection state.  The patching overrides the code I added here that consults the simulated state - so you get inconsistencies like e.g. can_connect returning true even though Container.pull will raise a ConnectionError.

sed-i commented:
> alertmanager tests won't work with this new behavior because they patch the get_system_info function

Thanks, I didn't notice! [Gone](https://github.com/canonical/alertmanager-k8s-operator/runs/5665341456?check_suite_focus=true) :)

PietroPasotti commented on ops/testing.py:
```diff
@@ -100,6 +123,23 @@ def __init__(
+    def set_can_connect(self, container: typing.Union[str, model.Container], val: bool):
+        """Change the simulated can_connect status of a container's underlying pebble client.
+
+        SIMULATE_CAN_CONNECT must be True before using this.
```
Must SIMULATE_CAN_CONNECT be set to True before calling this method, or before instantiating Harness? The top docstring seems to suggest the latter.

rwcarlsen replied:
This method will raise if it's not set to true when you call it.  *If* you are going to call set SIMULATE_CAN_CONNECT=true, then you should set it before creating the harness.  Clear?

PietroPasotti replied:
I'd say "Calling this method will raise an exception if testing.SIMULATE_CAN_CONNECT is not True."

rwcarlsen replied:
Agree - I pushed up a tweak.

PietroPasotti reviewed (approved):
LGTM

pengale reviewed (approved):
Per my comments elsewhere, I am +1 on this approach.
